packages/babel-core/src/config/partial.js

```diff
@@ -6,9 +6,43 @@ import { mergeOptions } from "./util";
 import { createItemFromDescriptor } from "./item";
 import { buildRootChain, type ConfigContext } from "./config-chain";
 import { getEnv } from "./helpers/environment";
-import { validate, type ValidatedOptions } from "./validation/options";
+import {
+  validate,
+  type ValidatedOptions,
+  type RootMode,
+} from "./validation/options";
+
+import { findConfigUpwards, type ConfigFile, type IgnoreFile } from "./files";
+
+function resolveRootMode(rootDir: string, rootMode: RootMode): string {
+  switch (rootMode) {
+    case "root":
+      return rootDir;
+
+    case "upward-optional": {
+      const upwardRootDir = findConfigUpwards(rootDir);
+      return upwardRootDir === null ? rootDir : upwardRootDir;
+    }
 
-import type { ConfigFile, IgnoreFile } from "./files";
+    case "upward": {
+      const upwardRootDir = findConfigUpwards(rootDir);
+      if (upwardRootDir !== null) return upwardRootDir;
+
+      throw Object.assign(
+        (new Error(
+          `Babel was run with rootMode:"${rootMode}" but a root could not ` +
```

Comment:
Nit: No need to interpolate, you can just write `rootMode: "upward"`.

Reply:
Thanks! I moved the code around and forgot to update this snippet.